chain/client/src/sync/epoch.rs

```diff
@@ -729,82 +773,156 @@ impl EpochSync {
+    ) -> Vec<ApprovalStake> {
+        let mut settlement = current_block_producers.to_vec();
+        let settlement_epoch_boundary = settlement.len();
+
+        settlement.extend_from_slice(next_block_producers);
+
+        let mut result = vec![];
+        let mut validators: HashMap<AccountId, usize> = HashMap::default();
+        for (ord, validator_stake) in settlement.into_iter().enumerate() {
+            let account_id = validator_stake.account_id();
+            match validators.get(account_id) {
+                None => {
+                    validators.insert(account_id.clone(), result.len());
+                    result
+                        .push(validator_stake.get_approval_stake(ord >= settlement_epoch_boundary));
+                }
+                Some(old_ord) => {
+                    if ord >= settlement_epoch_boundary {
+                        result[*old_ord].stake_next_epoch = validator_stake.stake();
+                    };
+                }
+            };
+        }
+        result
+    }
+
+    /// Verifies that EpochSyncProofPastEpochData's block_producers is valid
     fn verify_block_producer_handoff(
-        epoch: &EpochSyncProofPastEpochData,
+        epoch_index: usize,
+        block_producers: &Vec<ValidatorStake>,
         prev_epoch_protocol_version: ProtocolVersion,
         prev_epoch_next_bp_hash: &CryptoHash,
-        epoch_manager: &dyn EpochManagerAdapter,
     ) -> Result<(), Error> {
-        // Verify epoch.block_producers
         let bp_hash = Chain::compute_bp_hash_from_validator_stakes(
-            &epoch.block_producers,
+            block_producers,
             prev_epoch_protocol_version,
         )?;
         if bp_hash != *prev_epoch_next_bp_hash {
-            return Err(Error::InvalidEpochSyncProof("invalid block_producers".to_string()));
+            return Err(Error::InvalidEpochSyncProof(format!(
+                "invalid block_producers for epoch index {}",
+                epoch_index
+            )));
         }
+        Ok(())
+    }
 
-        // Verify epoch.last_final_block_header
+    /// Verifies that the epoch's last_final_block_header is correctly endorsed.
+    ///
+    /// Note that we do not trust next_epoch_block_producers. It doesn't matter what's given there;
+    /// it's just so that we can verify the signatures of the last final block of the epoch.
+    /// This block requires signatures from both this epoch's block producers and the next epoch's,
+    /// so we need to know both to verify the signatures; however, we only need this epoch's block
+    /// producers to be valid to verify that the signatures are valid. It is possible to to invent
+    /// an incorrect list (such as one with incorrect stake numbers) for the next epoch's block
+    /// producers so that we still pass the signature verification, but that's fine, as we still do
+    /// not trust the next epoch block producers; those will be verified later.
```

Comment:
First, I thought this is fine because we only need to verify signatures, and this is in sender' interest to provide correct public keys. But then I didn't understand what prevents sender to provide empty list of block producers, or set all stakes to zero so block will look approved, so next block producers will be spoiled.

After I looked broader, I saw that in the **next** iteration `next_epoch_block_producers` is compared with `next_bp_hash`. So if next epoch data was spoiled, this verification fails, because current epoch info can't be corrupted (induction?)

I wonder if the code or explanation could make it more clear, without looking what happens outside. It's unusual to use data which wasn't fully verified; while logic seems correct, audit becomes very time-consuming.

Reply:
So... the next epoch's block producers are just not trusted. We use them only to be able to figure out the endorsements' ordering so we can verify the final block against the current block producers.

When I think about this more, I think there's a way to simplify this. Instead of doing this complicated verification that looks half-correct, I can just include only the current epoch endorsements in the list of endorsements. And then I don't even need the next epoch's block producers anymore.

Same with protocol_version. We're not actually validating that. I'll turn that into a boolean that's still unverified but rely on the non-alias-ability of the two bp_hash formats to argue that the boolean doesn't need to be trusted.

Reply:
Thanks for raising this - this is very very helpful.